Add support for nrf52_dk bootloader

JammuKekkonen opened the pull request:
Bootloader for nrf52_dk is a intel hex file, which contains write instructions to the start of rom and also into registry which is outside rom region. This causes the current logic of forming the build regions to fail as it only supports bootloader that is in one segments which is at the start of rom.

This pr fixes bootloader build region generation in case where application size is restricted (space reserved for something else after application) and also adds support for nrf52_dk bootloader, which contains write instructions to registry which is outside rom region.

Also fixed printing of error case.

### Pull request type
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

Commit:
Fix bootloader build region generation and add support for bootloaders with multiple segments, as required by NRF52 bootloader.

JammuKekkonen commented:
@LiyouZhou 
@JuhaYlipekkala 
@yogpan01

yogpan01 commented:
Approved,
@ARMmbed/mbed-os-maintainers please review.

cmonr commented:
/morph build

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +425,21 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+            if len(part.segments()) == 1:
+                part_size = (part.maxaddr() - part.minaddr()) + 1
+            else:
+                # make same assumption as in region builder; first segments must fit.
```
I don't understand what this comment is about. Could you explain it?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -692,12 +692,13 @@ def _assign_new_offset(rom_start, start, new_offset, region_name):
     def _generate_bootloader_build(self, rom_start, rom_size):
         start = rom_start
         rom_end = rom_start + rom_size
+        max_app_addr = -1
```
Please don't use `-1` as a marker. Pythonic optional types are generally `None` or an instance of the type. Could you replace this `-1` with `None`?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +711,23 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+                # assume first segment is at start
+                # second at the end or outside ROM
+                # rest outside regular ROM
+                if part.segments()[0][0] != 0x0:
```
This is not always true; ROM does not always start at `0x0`. Please compare with `rom_start`.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +711,23 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+                part_size = part.segments()[0][1] - part.segments()[0][0]
+                part_size = Config._align_ceiling(rom_start + part_size, self.sectors) - rom_start
+                if part.segments()[1][0] < rom_end and self.target.restrict_size is None:
+                    if self.target.app_offset:
```
`target.app_offset` is not the only way that an application can be pushed further back: `target.header_format` and `target.header_offset` can also push an application further back. Either this check should check for all of them, or `max_app_addr` should be used unconditionally.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -722,9 +738,17 @@ def _generate_bootloader_build(self, rom_start, rom_size):
                     start, self.target.header_format)
                 yield region._replace(filename=self.target.header_format)
+
+            if max_app_addr != -1 and self.target.restrict_size is None and not self.target.app_offset:
```
Why does `target.app_offset` inhibit the use of `target.restrict_size` here?

cmonr commented:
Stopped build since changes have been requested.

Commit:
Review fixes for bootloader region generation

cmonr commented:
@theotherjimmy Mind doing a re-review?

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +425,21 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+                part_size = (part.maxaddr() - part.minaddr()) + 1
+            else:
+                # make same assumption as in region builder; first segments must fit.
+                part_size = part.segments()[0][1] - part.segments()[0][0]
```
Do you need the `+1` here?

theotherjimmy replied:
How is this branch different from the other branch when `part.segments() == 1`? Could we just use this implementation unconditionally?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +711,20 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+                # rest outside regular ROM
+                if part.segments()[0][0] != rom_start:
+                    raise ConfigException("bootloader segments not in order")
+                part_size = part.segments()[0][1] - part.segments()[0][0]
```
Size is off by 1 here as well.

Commit:
Review fixes 2 for bootloader region generation

JammuKekkonen commented:
PR updated according to feedback.

Tested with blinky.
K64F using existing bootloader
NRF52_DK with two different bootloaders; one that is located right after softdevice and one that is located at the end of flash (softdevice in beginning).

In both of the NRF52_DK tests, app header is located before app. @LiyouZhou.

theotherjimmy reviewed (approved):
Logic is much easier to follow, and a lot simpler. Thanks for making these changes.

cmonr commented:
/morph build

cmonr commented:
@JammuKekkonen Please take a look at the build errors.

simosillankorva commented:
The script seems to incorrecty assume that it could always find max address for the application from the same part of the parsed hex file. This causes the linking to fail in some build, as there is no space reserved for the application at all.

simosillankorva committed:
Remove max_app_addr checking from .hex

* Remove max_app_addr logic from tools/config/__init__.py as it was not working properly with some of the .hex files.

simosillankorva commented:
@cmonr Could you trigger the tests again to see if the lates change fixes the linking issues?

cmonr commented:
/morph build

bulislaw commented:
'''
13:47:46 Started by upstream project "mbed-os-build-matrix" build number 3284
13:47:46 originally caused by:
13:47:46  Started by upstream project "mbed-os-pipeline" build number 3490
13:47:46  originally caused by:
13:47:46   GitHub pull request #8097 of commit fac5ff45ed4953178a64e990d37b839c73ebadd3, no merge conflicts.
13:47:46 [EnvInject] - Loading node environment variables.
13:47:46 [EnvInject] - Preparing an environment for the build.
13:47:46 [EnvInject] - Keeping Jenkins system variables.
13:47:46 [EnvInject] - Keeping Jenkins build variables.
13:47:46 [EnvInject] - Injecting contributions.
13:47:46 Building remotely on ec2-linux-exporter (i-0cc56922f814f5da4) (linux_exporter linux_mbed_os_build ARM GCC_ARM) in workspace /builds/ws/mbed-os-build-matrix/3284/NRF52840_DK_ARM
13:47:46 [NRF52840_DK_ARM] $ env bash /tmp/jenkins4020254152608906915.sh
13:47:46 step : download sources...
13:47:46 build doesn't exists...
13:47:46 mkdir: cannot create directory ‘lockfile’: File exists
13:47:46 sources are been fetched by another process ...
13:47:46 waiting 100 sec for download to complete....
13:47:46 [NRF52840_DK_ARM] $ /bin/bash /tmp/jenkins4600830255462579272.sh
13:47:46 Extracting sources...
13:47:50 Extracting sources done..
13:47:50 [mbed] WARNING: Could not find mbed program in current path "/builds/ws/mbed-os-build-matrix/3284/NRF52840_DK_ARM/sources/mbed-os".
13:47:50 [mbed] WARNING: You can fix this by calling "mbed new ." in the root of your program.
13:47:50 ---
13:47:57 [mbed] WARNING: Unable to auto-install required Python modules.
13:47:57 ---
13:47:57 [mbed] WARNING: -----------------------------------------------------------------
13:47:57 [mbed] WARNING: The mbed OS tools in this program require the following Python modules: git, mbed_cloud_sdk, icetea
13:47:57 [mbed] WARNING: You can install all missing modules by running "pip install -r requirements.txt" in "/builds/ws/mbed-os-build-matrix/3284/NRF52840_DK_ARM/sources/mbed-os"
13:47:57 [mbed] WARNING: On Posix systems (Linux, Mac, etc) you might have to switch to superuser account or use "sudo"
13:47:57 ---
13:52:13 [mbed] ERROR: "/usr/bin/python" returned error code 1.
'''
Is that the error or are we hiding the real stuff somewhere? @0xc0170 @cmonr

bulislaw commented:
Nevermind, we are hiding the real failures...

cmonr commented:
@bulislaw No we're not: 
```
Merging Regions
  Filling region bootloader with /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/TARGET_SOFTDEVICE_S132_FULL/hex/s132_nrf52_5.0.0_softdevice.hex
  Filling region application with /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/./TESTS/mbed_hal/sleep_manager/sleep_manager_application.hex
Traceback (most recent call last):
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/test_api.py", line 2209, in build_test_worker
    bin_file, _ = build_project(*args, **kwargs)
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/build_api.py", line 548, in build_project
    merge_region_list(region_list, res, notify)
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/build_api.py", line 437, in merge_region_list
    merged.merge(part)
  File "/usr/local/lib/python2.7/dist-packages/intelhex/__init__.py", line 840, in merge
    'Data overlapped at address 0x%X' % i)
AddressOverlapError: Data overlapped at address 0x1000
Building project test_setup_failure (NRF52_DK, ARM)
```

cmonr commented:
Weird. Apparently there's a shortcut for closing PRs...

Error log: http://mbed-os.s3-eu-west-1.amazonaws.com/builds/8097/FAIL/NRF52_DK/ARM/fac5ff45ed4953178a64e990d37b839c73ebadd3_build_log_NRF52_DK_ARM.txt

@bulislaw You can find them by clicking the `Build artifacts/logs` link.

JanneKiiskila commented:
Seems we have some problems - who can help with this? @elm3 - can we get some support, please? 

```
Elf2Bin: callback_application
FromELF: /usr/local/ARM_Compiler_5.06u3/bin/fromelf --i32combined -o /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/./TESTS/mbed_functional/callback/callback_application.hex /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/./TESTS/mbed_functional/callback/callback_application.elf
[DEBUG] Return: 0
Merging Regions
  Filling region bootloader with /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/TARGET_SOFTDEVICE_S132_FULL/hex/s132_nrf52_5.0.0_softdevice.hex
  Filling region application with /builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/BUILD/tests/NRF52_DK/ARM/./TESTS/mbed_functional/callback/callback_application.hex
Traceback (most recent call last):
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/test_api.py", line 2209, in build_test_worker
    bin_file, _ = build_project(*args, **kwargs)
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/build_api.py", line 548, in build_project
    merge_region_list(region_list, res, notify)
  File "/builds/ws/mbed-os-build-matrix/3284/NRF52_DK_ARM/sources/mbed-os/tools/build_api.py", line 437, in merge_region_list
    merged.merge(part)
  File "/usr/local/lib/python2.7/dist-packages/intelhex/__init__.py", line 840, in merge
    'Data overlapped at address 0x%X' % i)
AddressOverlapError: Data overlapped at address 0x1000
```

juhoeskeli commented on tools/config/__init__.py:
```diff
@@ -710,8 +710,11 @@ def _generate_bootloader_build(self, rom_start, rom_size):
-            part_size = (part.maxaddr() - part.minaddr()) + 1
+
+            # segments returns start address and 'next after end' address
+            part_size = part.segments()[0][1] - part.segments()[0][0]
```
The logic needs to be extended here. s132_nrf52_5.0.0_softdevice.hex contains multiple segments that write into flash address space. It is not enough to do it like this, as the next segment from softdevice hex will collide with the start address we are extracting here.

0xc0170 replied:
Thanks @juhoeskeli for the review, @JammuKekkonen Please review

cmonr commented:
@JammuKekkonen Is there any progress being made on this?

simosillankorva commented:
@cmonr I think Jammu is no longer working on this issue. He told me that the implementation done here is lacking, but he was not able to figure out how to get it properly implemented. I wonder would it be possible to get someone who is familiar with this stuff to fix this issue?

naveenkaje commented:
@cmonr I am working on it IOTDEV-1675

Commit:
tools: process bootloader chunks

NRF Softdevice hex file can be in chunks.
Make sure we account for the space where the bootloader resides
by including all the chunks within the end of rom marker. This will
clearly mark out the initial bootloader region.

0xc0170 commented:
> @cmonr I am working on it IOTDEV-1675

Let us know once ready for review

naveenkaje commented:
posted https://github.com/ARMmbed/mbed-os/pull/8097/commits/2da8950b2a5e54ffba01dcdfb4762c0019b9259c Please review.

cmonr commented:
@theotherjimmy Please re-review.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +710,19 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+            # find the last valid address that's within rom_end and use that
+            # to compute the bootloader size
+            end_address = None
+            for each in part.segments():
```
```suggestion
            for start, stop in part.segments():
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +710,19 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+            # to compute the bootloader size
+            end_address = None
+            for each in part.segments():
+                if (each[1] < rom_end):
```
```suggestion
                if (stop < rom_end):
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +710,19 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+            end_address = None
+            for each in part.segments():
+                if (each[1] < rom_end):
+                    end_address = each[1]
```
```suggestion
                    end_address = stop
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -710,8 +710,19 @@ def _generate_bootloader_build(self, rom_start, rom_size):
+                    break
+            if end_address == None:
+                raise ConfigException("bootloader segments don't fit within rom region")
+            part_size = Config._align_ceiling(rom_start + (end_address - start), self.sectors) - rom_start
```
`start` == `rom_start`

```suggestion
            part_size = Config._align_ceiling(end_address, self.sectors) - rom_start
```

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +430,24 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+            # throw intelhex.AddressOverlapError if there's overlapping
+            part_size = 0
+            for es in part.segments():
+                part_size += es[1] - es[0]
```
I think this will compute `pad_size` incorrectly below, as this is the size is not something that makes sense when you have a hex file with more than one segment. The original code was written assuming that there was only ever one segment in a bootloder hex file.

theotherjimmy reviewed (changes requested):
I think the section that starts with `for es in part.segments()` implements the same logic as `merged.merge(part)`, and that change should probably be reverted.

Commit:
tools: update build_api and __init__ as per review comments

naveenkaje commented:
@theotherjimmy updated with https://github.com/ARMmbed/mbed-os/pull/8097/commits/e28c260763ddd52e3b882b993139bd0ea753eef6

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +430,33 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+                # Add padding in between segments starting from end of first segment
+                if (len(part.segments()) > 1 and (merged.maxaddr() != None)):
+                    pad_size = es[0] - (merged.maxaddr() + 1)
+                    merged.puts(merged.maxaddr()+1, padding * pad_size)
```
Please add spaces around `+` for consistency with the rest of the code base. Further, this should probably check to see if we _need_ padding, and this breaks when you have an image that's before the active region and after it, as it will add padding that overlaps with the active region.

theotherjimmy reviewed (changes requested):
I think the merging bit may unintentionally create an overlap under certain circumstances that we're designing this for.

JanneKiiskila commented:
@adbridge  @0xc0170  - can we get this to the next patch release, please.

adbridge commented:
@JanneKiiskila as 5.10.3 is being released today we will get this in for 5.10.4 in 2 weeks time .

JanneKiiskila commented:
@adbridge  - understood, @naveenkaje  - please fix review comments soonest.

adbridge commented:
@naveenkaje Can you please address @theotherjimmy last comment?

naveenkaje commented:
> @naveenkaje Can you please address @theotherjimmy last comment?

working on it

Commit:
tools: Update build_api padding method

Update the padding sequence so that the gaps
are padded after all the regions are merged.
This avoids overwriting active regions and
serializes the process.

0xc0170 commented:
Needs to be rereviewed

theotherjimmy commented on tools/build_api.py:
```diff
@@ -433,30 +433,25 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
-                merged.puts(merged.maxaddr() + 1, padding * pad_size)
+    # Hex file can have gaps, so no padding needed. While other formats may
+    # need padding. Iterate through segments and pad the gaps.
+    if (format != ".hex"):
```
Please remove the `(` and `)`, as they're not needed.

naveenkaje replied:
done

theotherjimmy commented on tools/build_api.py:
```diff
@@ -433,30 +433,25 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
-                if (len(part.segments()) > 1 and (merged.maxaddr() != None)):
-                    pad_size = es[0] - (merged.maxaddr() + 1)
-                    merged.puts(merged.maxaddr()+1, padding * pad_size)
                 part_size += es[1] - es[0]
```
Since we no longer use this, can we remove the computation of `part_size`?

naveenkaje replied:
part_size is used to check if we exceed region.size. Hence keeping it.

theotherjimmy replied:
But part_size does not correspond to any actual size here, it's the sum of the sizes of all of the included sectors. That's not a useful thing for comparing against region.size.

theotherjimmy commented on tools/build_api.py:
```diff
@@ -433,30 +433,25 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
-                    pad_size = es[0] - (merged.maxaddr() + 1)
-                    merged.puts(merged.maxaddr()+1, padding * pad_size)
                 part_size += es[1] - es[0]
                 merged.merge(part[es[0]:_end_addr_inclusive(es[1])])
```
Since this loop is no longer any different from `merged.merge(part)`, can we just do that?

theotherjimmy commented on tools/build_api.py:
```diff
@@ -433,30 +433,25 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+            if (begin < es[0]):
+                pad_size = es[0] - begin
+                merged.puts(begin, padding * pad_size)
+            begin = es[1] + 1
```
This loop is a bit hard to follow. How about:
```python
begin, _ = merged.segments()[0]
for start, stop in merged.segments()[1:]:
    pad_size = start - begin
    merged.puts(begin, padding * pad_size)
    begin = stop + 1
```

This shoud not block the PR, don't get hung up on it.

naveenkaje replied:
Thanks @theotherjimmy . I will make these changes and resubmit.

naveenkaje replied:
changed it.
We need to set begin = 0.
So that we pad any gaps if there are in the beginning.

Hence a slight change

```if format != ".hex":
        begin = 0
        for start, stop in merged.segments()[1:]:
            pad_size = start - begin
            merged.puts(begin, padding * pad_size)
            begin = stop + 1
```

Should I not assume that gaps can be at 0s?

or should we only fill gaps starting from the end of the first segment?

theotherjimmy replied:
@naveenkaje You cannot assume that ROM starts an 0, so you cannot start begin at 0. You could start it at the address of the first region instead.

Commit:
tools: Update build_api

Rename loop varibles for better readability and
formatting changes.

naveenkaje commented:
Updated with https://github.com/ARMmbed/mbed-os/pull/8097/commits/858a7a83216dc961b9aab249e638dd9339b88f5b

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +430,27 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+
+            # make same assumption as in region builder; first segment must fit.
+            # this is only to get a neat ToolException anyway, since IntelHex.merge will
+            # throw IntelHex.AddressOverlapError if there's overlapping
```
Can we remove this comment? It no longer describes the code at all.

theotherjimmy commented on tools/build_api.py:
```diff
@@ -401,6 +401,11 @@ def _fill_header(region_list, current_region):
         start += Config.header_member_size(member)
     return header
 
+def _end_addr_inclusive(addr):
```
Can we remove this too? I don't think it's used.

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +430,27 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+                part_size += stop - start
+            merged.merge(part)
+
             if part_size > region.size:
```
If we want to keep this check, we should use something better than part_size here. Perhaps we need to duplicate the "last segment before the end of ROM" finding that is used in config?

theotherjimmy commented on tools/build_api.py:
```diff
@@ -426,20 +430,27 @@ def merge_region_list(region_list, destination, notify, padding=b'\xFF'):
+    # Hex file can have gaps, so no padding needed. While other formats may
+    # need padding. Iterate through segments and pad the gaps.
+    if format != ".hex":
+        begin = 0
```
Probably mean start of the first region. This will break unmanaged bootloder mode, and managed bootloder mode when the first address of rom != 0.

naveenkaje replied:
Uploaded https://github.com/ARMmbed/mbed-os/pull/8097/commits/38ce7b13e7bbc7bf759d0eca76cd0d4d6076ce62

Commit:
tools: fix up build_api padding logic and clean up unused code

Patch from the end of the first segment. Remove unused code.

naveenkaje commented:
Change https://github.com/ARMmbed/mbed-os/pull/8097/commits/38ce7b13e7bbc7bf759d0eca76cd0d4d6076ce62 with changes based on discussion

cmonr commented:
@theotherjimmy Mind re-reviewing? Looks like comments have been addressed.

simosillankorva commented:
It seems like the current version is not working with the NRF52_DK bootloader build from mbed-bootloader-internal.

The bootloader seems to take most of the flash so the application is not fitting in:
```
Using ROM regions bootloader, header, application in this build.
  Region bootloader: size 0x79000, offset 0x0
  Region header: size 0x70, offset 0x79000
  Region application: size 0x6f90, offset 0x79070
```

Here are the regions compiled when with git version fac5ff45ed4953178a64e990d37b839c73ebadd3 from this branch that works fine with the bootloader:
```
Using ROM regions bootloader, header, application in this build.
  Region bootloader: size 0x1000, offset 0x0
  Region header: size 0x70, offset 0x1000
  Region application: size 0x7ef90, offset 0x1070
```

JanneKiiskila commented:
Size 0x79000 for the bootloader is just... It's just wrong. It should be less than 32 kB. Except we have 2 bootloaders, do we? The Nordic "mbr" and the mbed-bootloader, which one is supposed to go here?

naveenkaje commented:
@simosillankorva 

Could you please share how you are seeing that size for bootloader segments? I tried the following with MBR and SOFTDEVICE_NONE configurations
Please share the application/test that you have compiled and I will try the same.

MBR configuration
```
python -u ./tools/test.py -c -t GCC_ARM -m NRF52_DK --source . --build ./BUILD/tests/NRF52_DK/GCC_ARM --test-spec ./BUILD/tests/NRF52_DK/GCC_ARM/test_spec.json --build-data ./BUILD/tests/NRF52_DK/GCC_ARM/build_data.json --test-config /c/mbed-cli/client-lite/test_config.json -n tests-mbed_hal-sleep_manager --greentea
... <snip> ...
MBR configuration -> greentea test
Compile [ 99.9%]: rtc_api.c
Compile [100.0%]: us_ticker.c
Building project sleep_manager (NRF52_DK, GCC_ARM)
Scan: GCC_ARM
Scan: sleep_manager
Using ROM regions bootloader, application in this build.
  Region bootloader: size 0x1000, offset 0x0
  Region application: size 0x7f000, offset 0x1000
Compile [100.0%]: main.cpp
Link: sleep_manager_application
Elf2Bin: sleep_manager_application
Merging Regions
  Filling region bootloader with C:\mbed-cli\client-lite\mbed-os\BUILD\tests\NRF52_DK\GCC_ARM\targets\TARGET_NORDIC\TARGET_NRF5x\TARGET_SDK_14_2\TARGET_SOFTDEVICE_S132_MBR\hex/s132_nrf52_5.0.0_mbr.hex
  Filling region application with ./BUILD/tests/NRF52_DK/GCC_ARM\.\TESTS\mbed_hal\sleep_manager\sleep_manager_application.hex
Space used after regions merged: 0xe958
```

simosillankorva commented:
It seems like I had problem in my application configuration that caused the previous issue. But now I started to get this error: [ERROR] Can not place 'application' region inside previous region. when trying to build the mbed-cloud-client-example-internal from  https://github.com/ARMmbed/mbed-cloud-client-example-internal/pull/546.

naveenkaje commented:
Based on discussions with Simo, seems like in this configuration, the bootloader is built with

*Bootloader Build Logs and Details*
```
$mbed compile -m NRF52_DK -t GCC_ARM --profile tiny.json --app-config configs/nrf52_internal_flash_fake_rot.json
...
Restrict size None
Setting application space, starting at 475136, size 49152
Using ROM regions bootloader, application in this build.
_add_all_regions printing region list
[Region(name='bootloader', start=0, size=4096, active=False, filename=u'C:\\mbed-cli\\mbed-bootloader\\mbed-os\\targets\\TARGET_NORDIC\\TARGET_NRF5x\\TARGET_SDK_14_2\\TARGET_SOFTDEVICE_S132_MBR\\hex/s132_nrf52_5.0.0_mbr.hex'),
 Region(name='application', start=475136, size=49152, active=True, filename=None)]
  Region bootloader: size 0x1000, offset 0x0
  Region application: size 0xc000, offset 0x74000
Link: mbed-bootloader_application
Elf2Bin: mbed-bootloader_application
_generate_bootloader build rom_start 0 rom_size 524288
[(0, 2408)]
start 0 stop 2408
start 0 setting end address to 2408 stop 2408 rom_end 524288
setting start to 4096, increased by 4096
Restrict size None
Setting application space, starting at 475136, size 49152
Merging Regions
  Filling region bootloader with C:\mbed-cli\mbed-bootloader\mbed-os\targets\TARGET_NORDIC\TARGET_NRF5x\TARGET_SDK_14_2\TARGET_SOFTDEVICE_S132_MBR\hex/s132_nrf52_5.0.0_mbr.hex
  Filling region application with .\BUILD\NRF52_DK\GCC_ARM\mbed-bootloader_application.hex
Space used after regions merged: 0x78aec
Merging Regions
  Filling region application with .\BUILD\NRF52_DK\GCC_ARM\mbed-bootloader_application.hex
Space used after regions merged: 0x4aec
| Module                                      |     .text |   .data |      .bss |
|---------------------------------------------|-----------|---------|-----------|
| [fill]                                      |    26(+0) |   5(+0) |    25(+0) |
| [lib]\c_nano.a                              |   644(+0) | 100(+0) |    12(+0) |
| [lib]\gcc.a                                 |   832(+0) |   0(+0) |     0(+0) |
| [lib]\misc                                  |   252(+0) |  12(+0) |    28(+0) |
| mbed-cloud-client\update-client-hub         |  1888(+0) |   0(+0) |    28(+0) |
| mbed-os\drivers                             |   488(+0) |   0(+0) |     8(+0) |
| mbed-os\features                            |  1200(+0) |   4(+0) |     0(+0) |
| mbed-os\hal                                 |  1206(+0) |   5(+0) |    64(+0) |
| mbed-os\platform                            |   762(+0) |   4(+0) |   125(+0) |
| mbed-os\targets                             |  6950(+0) |  46(+0) |   505(+0) |
| minimal-printf\mbed_printf_implementation.o |   934(+0) |   0(+0) |     0(+0) |
| minimal-printf\mbed_printf_wrapper.o        |    32(+0) |   0(+0) |     0(+0) |
| source\active_application.o                 |  1395(+0) |   4(+0) |     8(+0) |
| source\bootloader_common.o                  |   246(+0) |   4(+0) | 16385(+0) |
| source\bootloader_platform.o                |    46(+0) |   0(+0) |     0(+0) |
| source\main.o                               |   715(+0) |   0(+0) |     0(+0) |
| source\upgrade.o                            |  1375(+0) |   0(+0) |     8(+0) |
| Subtotals                                   | 18991(+0) | 184(+0) | 17196(+0) |
Total Static RAM memory (data + bss): 17380(+0) bytes
Total Flash memory (text + data): 19175(+0) bytes
```

*Cloud Client Example Build logs and details*

```
$mbed compile -c -m NRF52_DK -t GCC_ARM --app-config configs/wifi_esp8266_v4.json --profile=release
...
Building project mbed-cloud-client-example-internal (NRF52_DK, GCC_ARM)
Scan: mbed-cloud-client-example-internal
Scan: env
bootloader file mbed-bootloader-nrf52_dk.hex
_generate_bootloader build rom_start 0 rom_size 524288
[(0, 2408), (4096, 140456), (143360, 210244)]
start 0 stop 2408
start 0 setting end address to 2408 stop 2408 rom_end 524288
start 4096 stop 140456
start 4096 setting end address to 140456 stop 140456 rom_end 524288
start 143360 stop 210244
start 143360 setting end address to 210244 stop 210244 rom_end 524288
going to try to place application region at offset 0x1000
[ERROR] Can not place 'application' region inside previous region
[mbed] Auto-installing missing Python modules...
[mbed] ERROR: "C:\Python27\python.exe" returned error code 1.
[mbed] ERROR: Command "C:\Python27\python.exe -u C:\mbed-cli\client-lite\nov14\example\mbed-cloud-client-example-internal\mbed-os\tools\make.py -t GCC_ARM -m NRF52_DK --profile release --source . --build .\BUILD\NRF52_DK\GCC_ARM -c --app-config configs/wifi_esp8266_v4.json" in "C:\mbed-cli\client-lite\nov14\example\mbed-cloud-client-example-internal"
```

It seems that the bootloader we generated with configs/nrf52_internal_flash_fake_rot.json has sections at 0x1000 and the specific target.app_offset=0x1000 is conflicting.

cmonr commented:
Making a note here. This is unlikely to make it in for 5.10.4. If a new commit is ready for testing, feel free to ping me and I can see what I can do to accelerate it though CI.

cmonr commented:
@simosillankorva I'm not too familiar with the cloud-client side of things, but I'm aware that we now have and are running tests against some portion of cloud-client (ref: https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci_cloud-client-test/).

Why do you need to set app_offset if the bootloader feature should be taking care of it automatically?

@JammuKekkonen @JanneKiiskila If y'all have any feedback with the current implementation, it would be helpful in determining if this current implementation still needs work.

Otherwise, I'm marking this as needs CI whenever a slot opens up.

cmonr commented:
/morph build

simosillankorva commented:
If you do no specify the app_offset it places the application like this:
```
> Using ROM regions bootloader, header, application in this build.
>   Region bootloader: size 0x79000, offset 0x0
>   Region header: size 0x70, offset 0x79000
>   Region application: size 0x6f90, offset 0x79070
```
If it would be automatically placed correctly then app_offset would not be needed.

0xc0170 commented:
@theotherjimmy Waiting for your approval for this

Retesting 

/morph test

theotherjimmy commented:
> If you do no specify the app_offset it places the application like this:

Because your bootloader contains sections that: 
 1) are in rom
 2) end within the erase block that ends at 0x79000

And managed bootloader mode assumes that your application comes after your bootloader.

0xc0170 commented:
Info: This PR has been re-bundled into a new rollup PR (#8800).

No further work is needed here, as once that PR is merged, this PR will also be closed and marked as merged.
If any more commits are made in this PR, this PR will remain open and have to go through CI on its own.

simosillankorva commented:
I am wondering is there some work ongoin still on this issue? If not what problem does it actually fix now?

0xc0170 commented:
@simosillankorva I believe this was completed and now just waiting for to be integrated via a rollup PR

JanneKiiskila commented:
@naveenkaje - how was it tested? I think our understanding still is that it doesn't work for us.

0xc0170 commented:
Removing this from rollup, adding label needs: work

JanneKiiskila commented:
Hold a second still @0xc0170 , @simosillankorva will still check the configs.

cmonr commented:
@JanneKiiskila Anything ever come from the checks?

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

0xc0170 commented:
Moved to 5.11.1. 

> Hold a second still @0xc0170 , @simosillankorva will still check the configs.

This still needs work as I understood.

cmonr commented:
@0xc0170 That sounds correct. Conversation has been taken offline.

cmonr commented:
My bad. Email conversation is still ongoing.

JanneKiiskila commented:
Internal ref: MBOTRIAGE-1871

0xc0170 commented:
Based on the internal conversation, it was decide that this patch will go in as it is and another patch for tools will come separately

@theotherjimmy @naveenkaje Please confirm

naveenkaje commented:
@0xc0170 Yes, that is correct. We plan to do a follow on patch to enable placing an application at an offset based on configuration parameters.

cmonr commented:
CI started
